src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -2250,28 +2250,60 @@ MergeTreeData::MutableDataPartPtr StorageReplicatedMergeTree::attachPartHelperFo
     if (detached_parts.empty())
         return {};
 
-    auto & detached_part_info = detached_parts.front();
+    for (auto & detached_part_info : detached_parts)
+    {
+        chassert(rename_parts.old_and_new_names.empty());
+        rename_parts.addPart(detached_part_info.dir_name, "attaching_" + detached_part_info.dir_name, detached_part_info.disk);
 
-    rename_parts.addPart(detached_part_info.dir_name, "attaching_" + detached_part_info.dir_name, detached_part_info.disk);
+        try
+        {
+            rename_parts.tryRenameAll();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, fmt::format("data race is possible when moving detached part, part {} is ignored", detached_part_info.dir_name));
+            rename_parts.rollBackAll();
+            continue;
+        }
 
-    try
-    {
-        rename_parts.tryRenameAll();
-    }
-    catch (...)
-    {
-        tryLogCurrentException(log, fmt::format("data race is possible when moving detached part, part {} is ignored", detached_part_info.dir_name));
-        return {};
-    }
+        const auto volume = std::make_shared<SingleDiskVolume>("volume_" + detached_part_info.dir_name, detached_part_info.disk);
+        auto part = getDataPartBuilder(entry.new_part_name, volume, fs::path(rename_parts.source_dir) / rename_parts.old_and_new_names.front().new_name, getReadSettings())
+            .withPartFormatFromDisk()
+            .build();
+
+        try
+        {
+            loadPartAndFixMetadataImpl(part, getContext());
+        }
+        catch (...)
```

Comment:
It's not necessarily data corruption, it could be a random disk read/write error?